design/scratchpad/Contract_upgrade.md

```diff
@@ -0,0 +1,170 @@
+
+### Version Tracking:
+* Each contract's version is now tracked in the NetworkConfig contract
+* The ContractVersion struct stores the name, version string, and implementation address
+* Clear history of upgrades and current versions
+
+### Upgrade Events:
+* ContractUpgraded event is emitted for individual upgrades
+* BatchUpgradeCompleted event is emitted for batch upgrades
+* Events include both old and new versions for tracking changes
+
+### Batch Upgrade Support:
+* The recordBatchUpgrade function allows upgrading multiple contracts in one transaction
+* Generates a unique hash for each batch upgrade
+* Maintains atomicity of upgrades
+
+```
+// 1. Deploy new implementation contracts
+CrossChain newCrossChainImpl = new CrossChain();
+DataAvailabilityRegistry newDARImpl = new DataAvailabilityRegistry();
+... 
+
+// 2. Upgrade the proxies
+proxyAdmin.upgradeAndCall(
+    crossChainProxy,
+    address(newCrossChainImpl),
+    abi.encodeWithSelector(newCrossChainImpl.initialize.selector, ...)
+);
+
+proxyAdmin.upgradeAndCall(
+    daRegistryProxy,
+    address(newDARImpl),
+    abi.encodeWithSelector(newDARImpl.initialize.selector, ...)
+);
+..
+
+// 3. Record the upgrades in NetworkConfig
+string[] memory names = new string[](2);
+string[] memory versions = new string[](2);
+address[] memory implementations = new address[](2);
+
+names[0] = "CrossChain";
+versions[0] = "2.0.0";
+implementations[0] = address(newCrossChainImpl);
+
+names[1] = "DataAvailabilityRegistry";
+versions[1] = "2.0.0";
+implementations[1] = address(newDARImpl);
+
+// This one we could use with Stefans batched safe tx model 
+networkConfig.recordBatchUpgrade(names, versions, implementations);
+```
+# Bridge Upgrade
+
+Given the old bridge isn't recognised as a system contract, we can just deploy the new version and when we upgrade the nodes it will get passed in. We can have some sort of ForkManager which will determine whether to use old bridge logic ie message bus or the new logic ()
+
+```
+// fork_manager.go
+package crosschain
+
+import (
+    "github.com/ethereum/go-ethereum/common"
+)
+
+// ForkConfig represents a hard fork configuration
+type ForkConfig struct {
+    Name        string
+    BlockNumber uint64
+    IsActive    bool
+    // Add any other fork-specific configurations here
+    UseNewBridge bool
+}
+
+// ForkManager handles fork transitions
+type ForkManager struct {
+    forks map[string]*ForkConfig
+    logger *log.Logger
+}
+
+func NewForkManager(logger *log.Logger) *ForkManager {
+    return &ForkManager{
+        forks: make(map[string]*ForkConfig),
+        logger: logger,
+    }
+}
+
+// RegisterFork registers a new fork configuration
+func (fm *ForkManager) RegisterFork(name string, blockNumber uint64, useNewBridge bool) {
```

Comment:
I think those must be hardcoded as it is usually in ethland. L1 contract emits upgraded to "hardfork-name" and then this manager does activateFork on that name or panics